backend/plugin/mfa/otp/otp.go

```diff
@@ -0,0 +1,101 @@
+// Package otp is the plugin for OTP Multi-Factor Authentication.
+package otp
+
+import (
+	"errors"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pquerna/otp/totp"
+	"golang.org/x/exp/slices"
+)
+
+const (
+	// issuerName is the name of the issuer of the OTP token.
+	issuerName = "Bytebase"
+	// secretMaxExpiredDuration is the maximum duration that a secret is valid.
+	// The secret is valid for 5 minutes.
+	secretMaxExpiredDuration = 5 * time.Minute
+)
+
+// TimeBasedReader is a reader that returns the same value for the same timestamp.
+type TimeBasedReader struct {
+	reader *strings.Reader
+}
+
+// NewTimeBasedReader creates a new TimeBasedReader for the given timestamp.
+func NewTimeBasedReader(timestamp time.Time) *TimeBasedReader {
+	formatedTimestampUnix := timestamp.Unix() - int64(timestamp.Second())
+	return &TimeBasedReader{
+		reader: strings.NewReader(strconv.FormatInt(formatedTimestampUnix, 10)),
+	}
+}
+
+func (r *TimeBasedReader) Read(p []byte) (int, error) {
+	return r.reader.Read(p)
+}
+
+// GenerateSecret generates a secret for the given account name and timestamp.
+func GenerateSecret(accountName string, timestamp time.Time) (string, error) {
+	key, err := totp.Generate(totp.GenerateOpts{
+		Issuer:      issuerName,
+		AccountName: accountName,
+		Rand:        NewTimeBasedReader(timestamp),
+	})
+	if err != nil {
+		return "", err
+	}
+	return key.Secret(), nil
+}
+
+// GetValidSecrets returns a list of valid secrets for the given account name and timestamp.
+func GetValidSecrets(accountName string, timestamp time.Time) ([]string, error) {
+	var secrets []string
```

Comment:
I don't know whether this could work. IIUC, we first need to normalize the timestamp to the minute boundary (e.g. 1 min 20 sec and 1 min 50 sec would all normalize to 1 min)

Reply:
Let's take `1 min 50 sec` as an example, The valid time range for submission is **[`1 min 50 sec`, `6 min 50 sec`)**. This matches our design.